### PR TITLE
Implement AdminActivityLogPageActionTest #6654

### LINKED ISSUE #6654: Implement AdminActivityLogPageActionTest
Currently, the `AdminActivityLogPageActionTest.java` is just a empty test case.

### PR DESCRIPTION
**Pull request by xpdavid**
Fixes #6654 

**Outline of Solution**

> 1. Fix several bugs in `AdminActivityLogPageAction.java` and `AdminActivityLogPageData.java` when writing test cases.
> 2. Using `LocalModulesServiceTestConfig` and `LocalLogService` to simulate LogService in GAE so that we can inject log data `typicalLogMessage.json`
> 3. Write comprehensive test cases for filter function in `AdminActivityLogPageAction` so that we can refactor it later.

### CODE REVIEW

**Comment by xpdavid** [comment]
@whipermr5 Reminder.

File: src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java [hunk 1]
```diff
@@ -74,13 +74,19 @@ private void setDefaultLogSearchPeriod() {
...
-    public void init(boolean ifShowAll, boolean ifShowTestData, List<ActivityLogEntry> logs) {
-        this.ifShowAll = ifShowAll;
-        this.ifShowTestData = ifShowTestData;
+    public void init(List<ActivityLogEntry> logs) {
         this.logs = logs;
 
     }
 
+    public void setIfShowAll(boolean val) {
+        ifShowAll = val;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java]: Can we rename this to shouldShowAll?

File: src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java [hunk 2]
```diff
@@ -74,13 +74,19 @@ private void setDefaultLogSearchPeriod() {
...
+    }
+
+    public void setIfShowTestData(boolean val) {
+        ifShowTestData = val;
```
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java]: Similarly, rename to shouldShowTestData

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 3]
```diff
@@ -1,27 +1,557 @@
...
+        // test statusMessage with filterQuery
+        String query = "person:idOfInstructor1OfCourse1";
+        action = getAction("filterQuery", query);
+        statusMessage = getShowPageResult(action).getStatusMessage();
+        verifyStatusMessage(statusMessage, 12, 1, yesterday);
+        // TODO: fix the bug
+        // Currently when the `person:xxx` like query is present,
+        // status message is supposed to contain the time in that person's timezone.
+        // But this function gets a huge bug: The query will be processed as lower case.
+        // The person google id contains upper case. And thus the information could be lost.
```
Comment (whipermr5) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Can Google IDs provided by Google login really contain upper case (I'm not sure; just asking)?

File: src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java (hunk 1 above)
Comment (xpdavid) [inline src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java]: Is the `setIfShowAll` more like a setter method? (although I admit that `ifShowAll` is a bad name)

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java (hunk 3 above)
Comment (xpdavid) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Google doesn't allow this. But our test data allows this :P

**Comment by wkurniawan07** [comment]
@xpdavid let us know again when ready for review. You can use the labels and review requests now.

File: src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java (hunk 1 above)
Comment (whipermr5) [inline src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java]: You'll be renaming in another PR?

Comment (xpdavid) [inline src/main/java/teammates/ui/pagedata/AdminActivityLogPageData.java]: Yes. I am going to refactor the AdminActivityLogAction after it has been covered with test cases.

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 4]
```diff
@@ -120,6 +129,51 @@ public void loginAsStudent(String userId) {
...
+     * @param nickname
+     * @param startTimeUsec
+     * @param endTimeUsec
+     * @param method
+     * @param resource
+     * @param httpVersion
+     * @param userAgent
+     * @param complete
+     * @param status
+     * @param referrer
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/GaeSimulation.java]: Clear all these parameters if they don't have any extra explanation.

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 5]
```diff
@@ -120,6 +129,51 @@ public void loginAsStudent(String userId) {
...
+                                       method, resource, httpVersion, userAgent, complete, status, referrer);
+    }
+
+    /**
+     * Adds a App log line in GAE.
+     *
+     * @param requestId
+     * @param time
+     * @param level
+     * @param message
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/GaeSimulation.java]: Similarly here.

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 6]
```diff
@@ -46,6 +49,8 @@
 
     protected LocalServiceTestHelper helper;
 
+    protected LocalLogService localLogService;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/driver/GaeSimulation.java]: This can be `private`, right?

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 7]
```diff
@@ -1,27 +1,564 @@
...
 
+    private static final String TYPICAL_LOG_MESSAGE = "/typicalLogMessage.json";
+
+    private static final int LOG_MESSAGE_INDEX_TODAY = 0;
+    private static final int LOG_MESSAGE_INDEX_YESTERDAY = 1;
+    private static final int LOG_MESSAGE_INDEX_TWO_DAYS_AGO = 2;
+
+    private static final int LOG_MESSAGE_INDEX_MANY_LOGS = 3;
+
+    private static final int LOG_MESSAGE_INTERVAL_MANY_LOGS = 130;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: You may want to explain the magic numbers that may not be so obvious, e.g. this one (why 130?).

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 8]
```diff
@@ -1,27 +1,564 @@
...
     protected void prepareTestData() {
-        // no test data used in this test
+        super.prepareTestData();
+        loadLogMessages();
+        initVariable();
+    }
+
+    private void loadLogMessages() {
+        try {
+            String pathToJsonFile = TestProperties.TEST_DATA_FOLDER + TYPICAL_LOG_MESSAGE;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Does `TYPICAL_LOG_MESSAGE` need to be a local variable? You can hard-code `/typicalLogMessage.json` here if only used once.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 9]
```diff
@@ -1,27 +1,564 @@
...
+        AjaxResult result = getAjaxResult(action);
+        AdminActivityLogPageData pageData = (AdminActivityLogPageData) result.data;
+        verifyStatusMessage(result.getStatusMessage(), totalLogs, filteredLogs, earliestDateInUtc);
+        verifyLogs(expected, pageData.getLogs());
+    }
+
+    private void verifyActionResult(int[][] expectedLogs, String... params) {
+        AdminActivityLogPageAction action = getAction(params);
+        ShowPageResult result = getShowPageResult(action);
+        AdminActivityLogPageData page = (AdminActivityLogPageData) result.data;
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: `page` -> `pageData`.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 10]
```diff
@@ -1,27 +1,566 @@
...
+import teammates.ui.controller.AjaxResult;
+import teammates.ui.controller.ShowPageResult;
+import teammates.ui.pagedata.AdminActivityLogPageData;
+import teammates.ui.pagedata.PageData;
+
+import com.google.gson.reflect.TypeToken;
 
 public class AdminActivityLogPageActionTest extends BaseActionTest {
 
+    // typicalLogMessage.json will give a List<List<String>>. These constant are indexes
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: No need to explicitly mention "typicalLogMessage.json". "The test data" will do just fine and more future-proof in case you rename the JSON file.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 11]
```diff
@@ -1,27 +1,566 @@
...
+        String statusMessage = getShowPageResult(action).getStatusMessage();
+        verifyStatusMessage(statusMessage, 12, 5, yesterday);
+        verifyLocalTimeInStatusMessage(statusMessage, yesterday, Const.SystemParams.ADMIN_TIME_ZONE_DOUBLE);
+
+        // test statusMessage with filterQuery
+        String query = "person:idOfInstructor1OfCourse1";
+        action = getAction("filterQuery", query);
+        statusMessage = getShowPageResult(action).getStatusMessage();
+        verifyStatusMessage(statusMessage, 12, 1, yesterday);
+        // TODO: fix the bug
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Do make the issue right after this is merged, otherwise we're adding more technical debts.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 12]
```diff
@@ -1,27 +1,566 @@
...
+        List<ActivityLogEntry> actualLogs = pageData.getLogs();
+        verifyLogs(expectedLogs, actualLogs);
+    }
+
+    /**
+     * Verifies actualLogs contains expectedLogs.
+     *
+     * <p>expectedLogs is a 2D array, the outer indices correspond to {@link #LOG_MESSAGE_INDEX_TODAY}
+     * {@link #LOG_MESSAGE_YESTDAY_INDEX} and {@link #LOG_MESSAGE_INDEX_TWO_DAYS_AGO}, the inner indices for
+     * every {@code LOG_MESSAGE_*_INDEX} correspond to the orders in {@link #TYPICAL_LOG_MESSAGE} file.
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: `TYPICAL_LOG_MESSAGE` file -> the test data. Same reason: future-proofing.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 13]
```diff
@@ -1,27 +1,566 @@
...
+     *
+     * @param expectedLogs Expected logs
+     * @param actualLogs Actual logs
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: These param descriptions look redundant.

**Comment by damithc** [comment]
Good work @xpdavid. 👍 Just waiting to see if the test code can be improved.

**Comment by damithc** [comment]
> Clarification for what? :P

Oops, my comment got lost. My question was whether those tests need to run in sequence. If not, we should break them into independent tests. The problem with sequential tests is that the n'th test depend on the state that was modified by (n-1) previous tests. That means the reader has to read all n tests to understand the n'th test.

**Comment by xpdavid** [comment]
> Oops, my comment got lost. My question was whether those tests need to run in sequence. If not, we should break them into independent tests. The problem with sequential tests is that the n'th test depend on the state that was modified by (n-1) previous tests. That means the reader has to read all n tests to understand the n'th test.

Actually, all of them are independent. So the test is like this. We manually inject the fake log messages into the LocalLogService and them query the log message.

**Comment by damithc** [comment]
> Actually, all of them are independent. So the test is like this. We manually inject the fake log messages into the LocalLogService and them query the log message.

That's good. Also consider using the proper test naming convention (`sut_scenario_expected`) if you break them into independent tests.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 14]
```diff
@@ -1,27 +1,551 @@
...
+        expected = new int[][]{{0, 1, 3, 4, 5, 7, 8, 9}};
+        verifyActionResult(expected, "all", "true");
+
+        // show everything
+        expected = new int[][]{{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}};
+        verifyActionResult(expected, "testdata", "true", "all", "true");
+    }
+
+    @Test(groups = "typicalLogs")
+    public void filterQuery_validQuery_querySuccessful() {
```
Comment (damithc) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Unless you have another case `filterQuery_validQuery_queryFails()` i.e. variations of the 3rd component, you can omit the 3rd component altogether. e.g. `filterQuery_validQuery()`

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 15]
```diff
@@ -1,27 +1,551 @@
...
+
+    private void verifyStatusMessage(String message, int totalLogs, int filteredLogs, Date earliestDateInUtc) {
+        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy HH");
+        sdf.setTimeZone(TimeZone.getTimeZone(Const.SystemParams.ADMIN_TIME_ZONE));
+
+        assertTrue(message.contains("Total Logs gone through in last search: " + totalLogs));
+        assertTrue(message.contains("Total Relevant Logs found in last search: " + filteredLogs));
+        assertTrue(message.contains("Logs are from following version(s): 1"));
+        assertTrue(message.contains("All available version(s): 1"));
+        // bug might be introduced when the time is 00:00 AM.
```
Comment (damithc) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Not sure how this comment is supposed to help the reader. May be provide some more info?

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 16]
```diff
@@ -1,27 +1,551 @@
...
+    private void verifyLocalTimeInStatusMessage(String message, Date timeInUtc, double localTimeZone) {
+        SimpleDateFormat sdf = new SimpleDateFormat("dd-MM-yyyy HH");
+        Calendar timeLocal = TimeHelper.convertToUserTimeZone(TimeHelper.dateToCalendar(timeInUtc), localTimeZone);
+        assertTrue(message.contains(sdf.format(timeLocal.getTime())));
+        assertTrue(message.contains(String.format("in Local Time Zone (%.1f)", localTimeZone)));
+    }
+
+    private void verifyLoadingLocalTimeAjaxResult(String expected, String role, String googleId, long timeInMillis) {
+        String[] params = new String[]{"logRole", role, "logGoogleId", googleId,
+                                       "logTimeInAdminTimeZone", String.valueOf(timeInMillis)};
```
Comment (damithc) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: Use 8-space indentation for line wrapping. We are moving away from custom indentation levels. But you can choose the line break location that gives the most readable code.

File: src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java [hunk 17]
```diff
@@ -1,27 +1,551 @@
...
+    }
+
+    private void insertLogMessagesAtTime(List<String> msgList, long timeMillis) {
+        insertLogMessageAtTimeWithInterval(msgList, timeMillis, 1);
+    }
+
+    private void insertLogMessageAtTimeWithInterval(List<String> msgList, long timeMillis, int intervalInSecond) {
+        int levelInfo = 1;
+        // bug might be introduced when the time is 00:00 AM.
+        // but this situation is really rare and can be solved by re-running the test case
```
Comment (damithc) [inline src/test/java/teammates/test/cases/action/AdminActivityLogPageActionTest.java]: If a phrase, start with a lower case letter and don't put a fullstop. If a full sentence, start with an upper case letter and end with a fullstop.
